pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/coreimpl/CachingShadowChainNode.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.symbols.table.coreimpl;
```

Comment:
the package name "coreimpl" sounds a bit unusual. That's the part, you made independent of pmd-java and could potentially be moved to core, right?
Maybe we should rename the sub-package with something like "shadow" or "nameshadow" or so... or maybe something with "hierarchy" in it?

Reply:
> That's the part, you made independent of pmd-java and could potentially be moved to core, right?

Indeed, which is why the packaging is a bit awkward... Maybe `chain` would be a better name?

ShadowChain is a bit awkward, but I wanted to avoid the name clash with the current Scope classes we have. I think in a world where Scope doesn't exist, ShadowChain could be renamed to ScopeChain, and placed eg in the package `nspmd.symbols.scopes`?